src/almond/worker_k8s.ts

```diff
@@ -0,0 +1,294 @@
+#!/usr/bin/env node
+// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of Almond
+//
+// Copyright 2021 The Board of Trustees of the Leland Stanford Junior University
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import express from 'express';
+import WebSocket from "ws";
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const wsjs = require('ws');
+
+import * as http from 'http';
+import * as rpc from 'transparent-rpc';
+import * as argparse from 'argparse';
+
+import PlatformModule, { PlatformOptions } from './platform';
+import JsonWebSocketAdapter from '../util/json_websocket';
+import * as i18n from '../util/i18n';
+import Engine from './engine';
+
+interface EngineState {
+    userId : number;
+    running : boolean;
+    sockets : Set<rpc.Socket>;
+    stopped : boolean;
+    engine ?: Engine;
+}
+
+class Worker {
+    private app : express.Application;
+    private server : http.Server;
+    private wss : WebSocket.Server;
+    private engines : Map<number, EngineState>; 
+    private stopped : boolean;
+
+    constructor(port : number) {
+        this.engines = new Map<number, EngineState>();
+        this.stopped = false;
+        this.app = express();
+        this.app.use(express.json());
+        this.app.use(express.urlencoded({ extended: true }));
+        this.app.set('port', port);
+
+        this.server = http.createServer(this.app);
+        this.server.on('upgrade', (request, socket, head) => {
+          const pathname = request.url;
+          if (pathname === '/engine') {
+              this.wss.handleUpgrade(request, socket, head, (ws) => {
+                  this.wss.emit('connection', ws, request);
+              });
+          } else {
+            console.log('Error http upgrade path: ' + pathname);
+            socket.destroy();
+          }
+        });
+
+        this.wss = new WebSocket.Server({ noServer: true });
+        this.wss.on('connection', async (ws : WebSocket, request : http.IncomingMessage) => {
+            this.connectWSEngine(ws);
+        });
+
+        this.app.post('/runEngine', (req, res, next) => {
```

Comment:
Btw it's a very minor point but I prefer kebab-case over camelCase for HTTP endpoints. It looks cleaner to me.